lib/taxonomy/govuk_taxonomy.rb

```diff
@@ -21,10 +21,39 @@ def all_taxons
       @_all_taxons ||= children.flat_map(&:tree) + draft_child_taxons.flat_map(&:tree)
     end
 
+    def matching_against_published_taxons(taxons)
+      @_published_taxons ||= matching_against_taxonomy_branches(taxons, children)
+    end
+
+    def matching_against_visible_draft_taxons(taxons)
+      @_draft_visible_taxons ||= matching_against_taxonomy_branches(taxons, draft_child_taxons)
+    end
+
   private
 
     def build_tree(taxon_hash)
       @tree_builder_class.new(taxon_hash).root_taxon
     end
+
+    def matching_against_taxonomy_branches(taxons, taxononomy_branches)
+      matching_taxons = []
+      taxononomy_branches.each do |branch|
+        matched_taxons = filter_against_taxonomy_branch(taxons, branch, [])
+        matching_taxons.concat(matched_taxons) if matched_taxons.any?
+      end
+      matching_taxons
```

Comment:
This looks like a good candidate for a `flat_map`, e.g.

```ruby
taxonomy_branches.flat_map do |branch|
  filter_against_taxonomy_branch(taxons, branch, [])
end
```

Reply:
Sounds good, will give it a go!